### PR TITLE
PRO-6660: modules order

### PR DESCRIPTION
**Pull request by myovchev**
## Summary

 Modules can now have a `before: "module-name"` property in their configuration to run (initialization) before another module. That works only for target modules that are initializable. 

## What are the specific steps to test this change?

```js
// my-module/index.js
module.exports = {
  // will be initialized before the core image piece
  before: '@apostrophecms/image'
}
```

## What kind of change does this PR introduce?
*(Check at least one)*

- [ ] Bug fix
- [x] New feature
- [ ] Refactor
- [ ] Documentation
- [ ] Build-related changes
- [ ] Other

## Make sure the PR fulfills these requirements:

- [x] It includes a) the existing issue ID being resolved, b) a convincing reason for adding this feature, or c) a clear description of the bug it resolves
- [x] The changelog is updated
- [ ] Related documentation has been updated
- [x] Related tests have been updated

If adding a new feature without an already open issue, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,8 +2,12 @@
 
 ## UNRELEASED
 
-### Fixes
+### Adds
 
+* Modules can now have a `before: "module-name"` property in their configuration to run (initialization) before another module. That works only for target modules that are initializable.
```
Comment (boutell) [inline CHANGELOG.md:7]: "Not initializable" means an "abstract base class" module, right? So the concept isn't relevant for them and it's OK to skip the warning I think.

File: index.js [hunk 2]
```diff
@@ -582,9 +582,92 @@ async function apostrophe(options, telemetry, rootSpan) {
...
+      };
+    }
+
+    // Loop through the modules that have a `before` property,
+    // validate and fill the initial `beforeSelf` metadata (first pass).
+    for (const name of beforeModules) {
+      const m = modules[name];
+      const before = m.before;
+      if (m.before === name) {
+        console.warn(`Module "${name}" has a 'before' property that references itself. Skipping.`);
```
Comment (boutell) [inline index.js:614]: Throw an error, stop the show. This is a new feature and we get to insist on good behavior.

File: index.js [hunk 3]
```diff
@@ -582,9 +582,92 @@ async function apostrophe(options, telemetry, rootSpan) {
...
+        continue;
+      }
+      if (!modules[before]) {
+        console.warn(`Module "${name}" has a 'before' property that references a non-existent module: "${before}". Skipping.`);
```
Comment (boutell) [inline index.js:618]: Throw an error, stop the show. This is a new feature and we get to insist on good behavior.

**Review by boutell — CHANGES_REQUESTED** [review]

File: index.js [hunk 4]
```diff
@@ -611,12 +611,10 @@ async function apostrophe(options, telemetry, rootSpan) {
       const m = modules[name];
       const before = m.before;
       if (m.before === name) {
-        console.warn(`Module "${name}" has a 'before' property that references itself. Skipping.`);
-        continue;
+        throw new Error(`Module "${name}" has a 'before' property that references itself. Skipping.`);
```
Comment (boutell) [inline index.js:614]: We're stopping, not skipping (:

Comment (myovchev) [reply index.js:614]: lol true :)

**Review by boutell — APPROVED** [review]

**Pull request merged**